src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionList.java

```diff
@@ -362,20 +362,31 @@ private static final class ParallelResumer {
                         nodes.put(n, se);
                     } else {
                         LOGGER.warning(() -> "Could not find FlowNode for " + se + " so it will not be resumed");
+                        // TODO: Should we call se.getContext().onFailure()?
                     }
                 } catch (IOException | InterruptedException x) {
                     LOGGER.log(Level.WARNING, "Could not look up FlowNode for " + se + " so it will not be resumed", x);
+                    // TODO: Should we call se.getContext().onFailure()?
                 }
             }
-            for (FlowNode n : nodes.keySet()) {
-                LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
-                scanner.setup(n);
-                for (FlowNode parent : scanner) {
-                    if (parent != n && nodes.containsKey(parent)) {
-                        enclosing.put(n, parent);
-                        break;
+            try {
+                for (FlowNode n : nodes.keySet()) {
+                    LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
+                    scanner.setup(n);
+                    for (FlowNode parent : scanner) {
+                        if (parent != n && nodes.containsKey(parent)) {
+                            enclosing.put(n, parent);
+                            break;
+                        }
                     }
                 }
+            } catch (Exception e) {
+                // TODO: Should we instead just try to resume steps with no respect to topological order?
```

Comment:
Not sure if this would be better or worse than killing all of the steps. If you end up here, something is pretty wrong with your flow graph. The main causes I can think of are if a `BlockEndNode.startId` points to a node that either doesn't exist, is not a `BlockStartNode`, or (the new case I just saw) points to a `BlockStartNode` with a newer ID than the end node, creating a cycle in the graph.

The question though is whether flow graph corruption necessarily means that trying to resume the steps is pointless, or if there is a decent chance that if we fall back to synchronous resumption of each step execution in order, the build might be able to complete successfully.

Reply:
Ah I was just looking for #221 the other day as a reference, because its original motivating use case is obsolete as of https://github.com/jenkinsci/ssh-agent-plugin/pull/152. Of course there might still be other steps which legitimately need context from an enclosing step in their `onResume`, but I am not aware of any offhand. So I think it would be reasonable to fall back to resuming all steps in random order. I would not expect the build to complete successfully, but it might be able to e.g. release external resources more gracefully.

Reply:
[c6b4fa9](https://github.com/jenkinsci/workflow-api-plugin/pull/349/commits/c6b4fa940e2289e124f88a1ffb929528b4ef202d) adjusts things so we resume all steps as long as they have a `FlowNode`. https://github.com/jenkinsci/workflow-api-plugin/pull/349#discussion_r1712183413 discusses the cases where the `FlowNode` is missing and/or corrupted.